plasmapy/tests/helpers/exceptions.py

```diff
@@ -0,0 +1,109 @@
+"""Exceptions that describe different types of test failures."""
+
+__all__ = [
+    "Failed",
+    "UnexpectedResultError",
+    "InconsistentTypeError",
+    "MissingExceptionError",
+    "UnexpectedExceptionError",
+    "MissingExceptionError",
+    "UnexpectedWarningError",
+    "MissingWarningError",
+    "InvalidTestError",
+    "ExceptionMismatchError",
+    "WarningMismatchError",
+]
+
+import pytest
+
+# TODO: get the error message while raising this to be simply "Failed"
+#       rather than plasmapy.tests.helpers.exceptions.Failed
+
+
+class Failed(pytest.fail.Exception):
+    """
+    Base exception for test failures.
+
+    Notes
+    -----
+    This exception was derived from `~pytest.fail.Exception`, which in
+    turn was derived from `BaseException` (not `Exception`).
+    """
+
+    pass
+
+
+class InconsistentTypeError(Failed):
+    """
+    Exception for when the type of the actual result differs from the
+    type of the expected result.
+    """
+
+    pass
+
+
+class MissingExceptionError(Failed):
+    """
+    Exception for when the expected exception is not raised.
+    """
+
+    pass
+
+
+class UnexpectedExceptionError(Failed):
+    """
+    Exception for when an exception is raised unexpectedly.
+    """
+
+    pass
+
+
+class ExceptionMismatchError(UnexpectedExceptionError, MissingExceptionError):
+    """
+    Exception for when an exception is expected, but a different
+    exception is raised.
+    """
+
+    pass
+
```

Comment:
This feels a little to "fine" detail to me.  Why not have a `RaisesError` for any wrong `Exception` type failures?  The description string will do all the necessary identification of mismatches, unexpected, etc.

Reply:
There are a couple of reasons why I chose to have a different exception for each of the different types of test failures.
 - I want the error messages to be as informative as possible, as quickly as possible.  I chose the names of these exceptions to be highly descriptive so that developers can get a good idea of what the problem is just by reading the name of the exception.  As someone who has a bad habit of skimming through error messages without actually reading them, a descriptive name really does help.
 - Having more specific exceptions allows more specific exception handling.  There are some possibilities for extending additional test runners (e.g., to replace `run_test_equivalent_calls`) that may need to handle different exceptions differently.

Reply:
I'm still not sold on this, but I won't argue it since I see the reasoning.  With that said, they should probably all inherit from a `RaisesError` exception, then you can catch all variations in the try-statement.